proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1113 @@
+can be executed at the later runtime stage. In practice, most programs can be
+divided into small portions that are executed at compile-time (e.g. the
+definition of ``makeLenses``), and the majority of the code that is not.
+However, there are currently no clear boundaries between stages, so the compiler
+must pessimistically assume that anything may be needed at compile-time or
+runtime.
+
+This has a variety of negative consequences. In particular:
+
+* Using Template Haskell causes compile-time performance to suffer due to
+  unnecessary (re)compilation.  This is particularly relevant for interactive
+  use of the compiler within an IDE (e.g. via Haskell Language Server).
+
+* Cross-compilation is made more difficult by the need to compile code for a
+  platform even though it will never be executed on that platform.
+
+*Levels* are the mechanism which the typechecker uses to ensure that staged evaluation
+is possible. Certain program contexts require identifiers to be at specific levels.
+In order to fix level-incorrect programs, programmers rely on implicit *cross-stage persistence* (CSP): the assumption that if
+a definition can be used at one level, then it can also be used at a later level. Cross-stage persistence is the root
+cause of the unnecessary compilation: if a module is used at compile time (e.g. as a dependency
+of a module with ``TemplateHaskell`` enabled),
+the compiler must assume it may be needed for runtime as well.
+
+Thus we propose a new
+pair of extensions that allow programmers to write level-correct programs
+without ubiquitous cross-stage persistence:
+
+* ``NoImplicitStagePersistence`` forbids normal top-level identifiers from
+  occurring within top-level splices or quotes, and
+
+* ``ExplicitLevelImports`` allows imports that explicitly enable the use of the
+  imported identifiers within top-level splices or quotes.
+
+This proposal draws on ideas discussed previously in
+`proposal #243: Stage Hygiene for Template Haskell
+<https://github.com/ghc-proposals/ghc-proposals/pull/243>`_ and
+`proposal #412: Explicit Splice Imports
+<https://github.com/ghc-proposals/ghc-proposals/pull/412>`_.
+
+
+Motivation
+----------
+
+Level-correct programs are necessary when using staged programming so
+that the program can be cleanly separated into compile-time and runtime
+portions. The existing mechanism to ensure level-correctness for imported
+identifiers is called *path-based cross stage persistence*: informally, it allows you to
+use imported identifiers at any level.
+We want to explicitly control this, because it leads to the need to compile all modules
+in a project for both runtime and compile time.
+
+This proposal introduces an explicit means to control the level at which identifiers
+are imported at. Therefore instead of relying on implicit persistence of an imported
+identifier, the programmer has to explicitly request for the identifier to be available
+at a later or earlier level.
```

Comment:
I'm waking up, this morning, with a question. Are both directions of implicit persistence (the ability to be used in later levels and the ability to be used in earlier levels) equally problematic? What parts of the motivations would be lost if we allowed later-levels but not earlier?

Reply:
If you allow later moving things to later levels but not earlier levels then a module `M` needs to demand itself and all it's dependencies for all future stages. So it isn't quite as bad as the possibility of moving things to negative levels as well but it didn't seem worth splitting downs to even more fine-grained specification.

Reply:
Sure, but what are concrete consequences? Do you have an example (even far fetched) of a build that would go slower?

Reply:
Still allowing positive implicit persistence in `NoImplicitStagePersistence` means we might compile modules for runtime when we shouldn't need to.

Example:

```hs
module Help where
  foo = 1
{-# LANGUAGE TemplateHaskellQuotes, ExplicitLevelImports #-}
module A where 
  import quote Help
  bar = [| foo |]
{-# LANGUAGE TemplateHaskell, ExplicitLevelImports #-}
module B where
  import splice A
  foo = $(if foo == 1 then [| True |] else [| False |])
```

Assuming the relaxed spec for `NoImplicitStagePersistence` means that 

* `A` imports `Help` at level 1 and higher, and `B` imports `A` at level -1 and higher,
* so requiring `B @ R` means we require `A @ C`, `Help @ R`, **but also `A @ R`**. The latter is redundant.

Assuming the proposed spec for `NoImplicitStagePersistence` means that 

* `A` imports `Help` at level 1, and `B` imports `A` at level -1,
* so requiring `B @ R` means we require `A @ C` and `Help @ R`, **but not `A @ R`**.

Reply:
Note that Arnaud's "relaxed `NoImplicitStagePersistence`" semantics is basically what happens under my "sharper spec" for `ImplicitStageRestriction`+`TemplateHaskellQuotes` at the end of https://github.com/ghc-proposals/ghc-proposals/pull/682#pullrequestreview-2448500943. This particular case simplifies to

> A module `M` enabling `ImplicitStagePersistence` imports its dependencies at level `min(0, explicit import level)` and higher if `TemplateHaskellQuotes` is enabled.
> Furthermore, if `M` enables `ImplicitStagePersistence` and `TemplateHaskellQuotes` is enabled, then any import of `M` is an import at level `min(0, explicit import level)` and higher.

Reply:
Ah, yes, it's a cross-compilation thing, right? Level -1 is potentially built for a different target than level 0 and more. So persistence from level -1 to later levels potentially causes extra compilation. But not level 0 and later. Maybe the reasoning behind the design is that it's probably a bad idea to special case -1, so instead we just disable persistence all together.

@mpickering I know this is a little bit of a chore, but it'd be really nice if these things (counter example, and reasoning behind this choice (the real reasoning not necessarily my musings)) were documented in the proposal.

Reply:
Level 0 to 1 can also be for different platforms. Think about quoting some OS-specific IO code.

Reply:
@Ericson2314 sorry I don't see it, can you enlighten me?

Reply:
> Ah, yes, it's a cross-compilation thing, right?

Do you mean that positive stage persistence (your proposed relaxed spec of `NoImplicitStagePersistence`) would be functionally indistinguishable (i.e. in terms of build parallelism) from the proposed one if it weren't for cross-compilation? 

Then yes, I think you are right (modulo perhaps a level vs. stage confusion), for the following reason: If `host=target`, then depending on `M` at runtime stage `R` is (strictly) weaker than depending on `M @ C`. Example (without `ExplicitLevelImports`):

```hs
module M where
  fun x = x + 1
  spliceFun x = [| $(lift x) + 1 |]

{-# LANGUAGE TemplateHaskell #-}
module D where
  import M
  f x | fun x == 42 = 0
      | otherwise   = $(spliceFun 42)
```

`D @ R` needs `M.spliceFun` as compiled object code (so `M @ C`).
It doesn't affect build parallelism if code in `D` also calls `M.fun` (that would need `M @ R`). Either way you need to wait for `M` to be compiled. 
So "importing on level n" is the same as "importing on level n and higher" in a non-cross setting. 

However, if `host /= target` then `@ C` denotes a different build output than `@ R`, so it matters to be precise in "importing on level n".

For the same reason, the combination `TemplateHaskellQuotes` + `ImplicitStagePersistence` should be pretty harmless for build parallelism today (and is probably the reason for its existence in the first place!), in contrast to `TemplateHaskell` + `ImplicitStagePersistence`.

So it appears that cross-compilation *is* part of the motivation for the proposed `NoImplicitStagePersistence`; otherwise we could go with Arnaud's relaxed variant as well, as far as I can see.

It appears to me that `(No)ImplicitStagePersistence` and its interaction with and without `TemplateHaskell` or `TemplateHaskellQuotes` is far more subtle than `ExplicitLevelImports`, yet still underspecified in the proposal.

Reply:
@sgraf812 Even without cross compilation you might compile in a different way for `C` vs `R`. For example, you may compile `C` dynamic way but `R` is profiled way. Or `C` is static way, `R` is "no-code" way.

Reply:
Aha, that's great insight! I assume "C is static way, R is "no-code" way" is used by HLS? It would be good to persist this kind of knowledge somewhere, although I feat that this GHC proposal might be the wrong place.

What do you think about the sharper spec in https://github.com/ghc-proposals/ghc-proposals/pull/682#pullrequestreview-2448500943?

Reply:
```haskell
myIOQuote = 
#if mingw32_TARGET_OS
   [| windowsIO foo bar |]
#else
   [| unixIO foo bar |]
#endif
```
note the use of `TARGET`, as opposed to `HOST` or  `BUILD`. Each of those corresponds to one of the 3 module-local levels.

Reply:
@Ericson2314 I'm afraid you are way ahead of me there. I still don't understand what you have in mind. That being said, @mpickering 's example, just above yours, convinced me.

Reply:
@aspiwack My point is about matters that we can ignore for now per @mpickering saying

> I think this part is getting more into how levels are interpreted as stages, something which the proposal is trying to avoid.

Just in case anyone is curious (this is notes on the future implementation MR, not the proposal! :)):

What I am saying is we do have 2 stages in the sense of "do I need interface or also implementation", but at least 3 stages in the same of "how many different variations of the module do I end up needing. This is because while the interface should ideally only depend on the level 0 platform, the implementation also depends on the the level 1 platform because we are quoting an interface which can vary based on platform.

Thankfully, we can be sure that the level -1 platform still doesn't matter: we're doing all the compilation on one machine, and if that's not the case thanks to cached builds, it is tracked via ABI hashes. If someone wants to carefully `ghc -M` such that we have a heterogeneous build with different GHCs running on different build platforms (do some compilation on windows, hand of hi files, do some compilation on Unix) etc. that is allowed and formally still correct (the difference in ABI hashes do not indicate a difference in intent of what we are supposed to end up with).